### ISSUE TITLE
How to disable loud speaker in video call

### ISSUE THREAD

**Issue by rashidabbas** [twilio/video-quickstart-ios#379]
I am using video call in my project but I am facing issue when participants are connected into room then call is on loudspeaker not voice. 

kindly guide me how can I disable loudspeaker in swift iOS.

**Comment by piyushtank** [comment]
@rashidabbas By default the audio routes to speaker, however you can overwrite it. 

The following example demonstrates changing the audio route from speaker to receiver:
(You can find example snippets in `TVIDefaultAudioDevice.h` header)

```
      TVODefaultAudioDevice *audioDevice = [TVODefaultAudioDevice audioDevice];

      TwilioVoice.audioDevice = audioDevice;

      //...connect to a Call with audioDevice. By default the audio route will be configured to speaker.

      audioDevice.block =  ^ {
          // We will execute `kTVODefaultAVAudioSessionConfigurationBlock` first.
          kTVODefaultAVAudioSessionConfigurationBlock();

          // Overwrite the audio route
          AVAudioSession *session = [AVAudioSession sharedInstance];
          NSError *error = nil;
          if (![session setMode:AVAudioSessionModeVoiceChat error:&error]) {
              NSLog(@"AVAudiosession setMode %@",error);
          }

          if (![session overrideOutputAudioPort:AVAudioSessionPortOverrideNone error:&error]) {
              NSLog(@"AVAudiosession overrideOutputAudioPort %@",error);
          }
      };
      audioDevice.block();
  
     //  Build connect options of your choice
     TVIConnectOptions *connectOptions = 

     TVIRoom *room = [TwilioVideo connectWithOptions:connectOptions];
```
Let me know if you have any questions.

Best,
Piyush

**Comment by rashidabbas** [comment]
@piyushtank thanks for response but I am using swift into my project so can you please give me in swift syntax because I don't know about objective c.

And this piece of code should be before connecting to room or after connecting with room?

Thanks

**Comment by piyushtank** [comment]
@rashidabbas You should assign `TwilioVideo.audioDevice` prior to creating any media track. If you want the audio routed to receiver always, I would execute the code snippet at the very beginning. 
```
        // TwilioVideo.audioDevice must be assigned prior to any audio or video track creation.
        let audioDevice = TVIDefaultAudioDevice()
        TwilioVideo.audioDevice = audioDevice

        // By default Audio is roouted to speaker. Configure audio session in the block.
        audioDevice.block = {
            // We will execute `kTVODefaultAVAudioSessionConfigurationBlock` first.
            kDefaultAVAudioSessionConfigurationBlock()

            do {
                try AVAudioSession.sharedInstance().setMode(.voiceChat)
                try AVAudioSession.sharedInstance().overrideOutputAudioPort(.none)
            } catch {
                print(error)
            }
        }

        // Execute the block to apply settings on AVAudioSession
        audioDevice.block();
```

While in a call, if you wish to change the audio route to speaker, set mode to `.videoChat` instead of `voiceChat` in the block above and execute the block. 

Let me know if you have any questions.

**Comment by irfanelahi** [comment]
Hi @piyushtank ,
I am stuck at changing audio route while in a call. I want user to toggle between audio routes on button click. When I try to change the audio route on runtime by executing the same code above, app crashes.

Error goes like this:
WebRTC does not allow updating the audio device once the media stack is created...

Can you please tell me what am I missing here.

Best,
Irfan

**Comment by piyushtank** [comment]
Hi @irfanelahi 

You should set the `TwilioVideo.audioDevice` only once. You should not set `TwilioVideo.audioDevice` every time you change the audio route.

Let me know if this fixes the problem.

Best,
Piyush

**Comment by piyushtank** [comment]
Closing the ticket assuming the problem is fixed at your end. Feel free to re-open if you run into the issue again.

**Comment by patelvivek9115** [comment]
@piyushtank Applying this code in my viewDidLoad(), when I pop the view controller and coming back I am getting crash 

> Terminating app due to uncaught exception 'TVIAudioDeviceMustBeSetBeforeMediaStackSetup', reason: 'WebRTC does not allow updating the audio device once the media stack is created (i.e once a track is created or memory allocated for previously created room or connect option is not deleted). You should provide the audio device to SDK prior to creating the tracks.'

```
TwilioVideoSDK.audioDevice = audioDevice
self.audioDevice.block = {
                do {
                    
                    // kDefaultAVAudioSessionConfigurationBlock()
                    
                    let audioSession = AVAudioSession.sharedInstance()
                    try audioSession.setMode(.voiceChat)
                } catch let error as NSError {
                    print("Fail: \(error.localizedDescription)")
                }
            }
            self.audioDevice.block()
```

**Comment by Cyangineer** [comment]
After awhile of searching and trial and error I figured out my problem on how to use the voice speaker on video call and also toggling back and forth from loudspeaker to voice speaker.

I have a function called `startPreview()`. That's where I _first init_ a track which is the `localVideoTrack`. Once you find where your first track init is placed `TwilioVideo.audioDevice = audioDevice` before that. That got rid of the error. Placement of that line is critical.

Then I put this block of code right after.

```
        self.audioDevice.block = {
            do {
                
                let audioSession = AVAudioSession.sharedInstance()
                try audioSession.setMode(AVAudioSessionModeVoiceChat)
                
            } catch {
                print("Fail: \(error.localizedDescription)")
            }
        }
        
        self.audioDevice.block()
```

And that allowed the audio to go through the voice speaker.

Now to toggle the loudspeaker on and off I only put `TwilioVideo.audioDevice = audioDevice` to the beginning of `startPreview()` and put the block in a UISwitch IBAction like this:

```
@IBAction func toggleSwitch(_ sender: UISwitch) {
        self.audioDevice.block = {
            do {
                
                let audioSession = AVAudioSession.sharedInstance()
                
                if(sender.isOn) {
                    try audioSession.setMode(AVAudioSessionModeVideoChat)
                } else {
                    try audioSession.setMode(AVAudioSessionModeVoiceChat)
                }
                
            } catch {
                print("Fail: \(error.localizedDescription)")
            }
        }
        
        self.audioDevice.block()
    }
```

I hope this helps. I also made a [gist](https://gist.github.com/Cyangineer/aa204ba3025c2d1cce430aec9cb9f0e5) for easier access.

**Comment by ALFriedman** [comment]
We are using Swift, and this same issue is baffling me. I tried following the answers in this thread as well as the gist and various other examples. Some of the answers include this in the block:

Some examples include "DefaultAudioDevice.DefaultAVAudioSessionConfigurationBlock()"  (or the older variant)
I tried with that line and without. No change.

Some of the examples call "audioSession.setActive(true)". 
I tried with and without. No change.

I also tried placing the initial audioDevice block in various locations. One thing I noticed in debugging is that it fails with "AVAudioSessionErrorCodeIsBusy" when calling "session setAggregatedIOPreference:AVAudioSessionIOTypeAggregated" inside the "DefaultAudioDevice.DefaultAVAudioSessionConfigurationBlock()" method. I'm not sure if that matters or not.

I even tried having two different blocks in different places to initialize (which I know I shouldn't need). One in the viewDidLoad method, just after "TwilioVideoSDK.audioDevice = audioDevice" and one later just before "prepareLocalMedia" and calling "TwilioVideoSDK.connect". 

In debugging, I noticed that the "audioDevice.block()" is being called when I call it manually, but ALSO when I call "TwilioVideoSDK.connect".

The first block, called in viewDidLoad looks like this:

`let audioSession = AVAudioSession.sharedInstance()
DefaultAudioDevice.DefaultAVAudioSessionConfigurationBlock()
LogInfo("Twilio Audio category \(audioSession.category)")
LogInfo("Twilio Audio mode \(audioSession.mode)")
try audioSession.setMode(AVAudioSessionModeVoiceChat)
try audioSession.overrideOutputAudioPort(AVAudioSession.PortOverride.none)`

And prints:
Twilio Audio category AVAudioSessionCategoryPlayAndRecord
Twilio Audio mode AVAudioSessionModeVideoChat

I know I shouldn't need multiple blocks, but I was playing with various options and placement. So right now I have another block just before setting up local audio and video tracks and connecting to a room:

`let audioSession = AVAudioSession.sharedInstance()
LogInfo("Twilio Audio 2 category \(audioSession.category)")
LogInfo("Twilio Audio 2 mode \(audioSession.mode)")
try audioSession.setMode(AVAudioSessionModeVoiceChat)
try audioSession.overrideOutputAudioPort(AVAudioSession.PortOverride.none)
try audioSession.setActive(true)
LogInfo("Twilio Audio 3 category \(audioSession.category)")
LogInfo("Twilio Audio 3 mode \(audioSession.mode)")`


And print the following as output: (**Note**: Where is it getting "AVAudioSessionModeVideoRecording" from?)
Twilio Audio 2 category AVAudioSessionCategoryPlayAndRecord
Twilio Audio 2 mode AVAudioSessionModeVideoRecording

Twilio Audio 3 category AVAudioSessionCategoryPlayAndRecord
Twilio Audio 3 mode AVAudioSessionModeVoiceChat

And then when calling "TwilioVideoSDK.connect" the block gets called again for some reason and prints
Twilio Audio 2 category AVAudioSessionCategoryPlayAndRecord
Twilio Audio 2 mode AVAudioSessionModeVoiceChat

Twilio Audio 3 category AVAudioSessionCategoryPlayAndRecord
Twilio Audio 3 mode AVAudioSessionModeVoiceChat

**Questions**: If the block gets called when connecting, should I also be calling it manually "self.audioDevice.block()"? I've seen various placements for calling that block in examples. @Cyangineer it looks like you call it just after initializing local media but just before calling "connect", right?

I do also have a speaker button that does the same as @Cyangineer 's toggle switch. It's getting called, but audio still comes through the speaker.

Any ideas of why I can't turn off speaker?
